src/Lean/Elab/Tactic/BVDecide/LRAT/Internal/Clause.lean

```diff
@@ -0,0 +1,418 @@
+  empty : β
+  empty_eq : toList empty = []
+  unit : Literal α → β
+  unit_eq : ∀ l : Literal α, toList (unit l) = [l]
+  isUnit : β → Option (Literal α)
+  isUnit_iff : ∀ c : β, ∀ l : Literal α, isUnit c = some l ↔ toList c = [l]
+  negate : β → CNF.Clause α
+  negate_eq : ∀ c : β, negate c = (toList c).map Literal.negate
+  /-- Returns none if the result is a tautology. -/
+  insert : β → Literal α → Option β
+  delete : β → Literal α → β
+  delete_iff : ∀ c : β, ∀ l : Literal α, ∀ l' : Literal α,
+    l' ∈ toList (delete c l) ↔ l' ≠ l ∧ l' ∈ toList c
+  contains : β → Literal α → Bool
+  contains_iff : ∀ c : β, ∀ l : Literal α, contains c l ↔ l ∈ toList c
+  /-- Reduces the clause with respect to the given assignment -/
+  reduce : β → Array Assignment → ReduceResult α
+
+namespace Clause
+
+instance : Entails α (Literal α) where
+  eval := fun p l => p l.1 = l.2
+
+instance (p : α → Bool) (l : Literal α) : Decidable (p ⊨ l) :=
+  inferInstanceAs (Decidable (p l.1 = l.2))
+
+def eval [Clause α β] (a : α → Bool) (c : β) : Bool :=
+  (toList c).any fun (l : Literal α) => a ⊨ l
+
+instance [Clause α β] : Entails α β where
+  eval a c := Clause.eval a c
+
+instance [Clause α β] (p : α → Bool) (c : β) : Decidable (p ⊨ c) :=
+  inferInstanceAs (Decidable (Clause.eval p c = true))
+
+instance [Clause α β] : Inhabited β where
+  default := empty
+
+end Clause
+
+/--
+The `DefaultClause` structure is primarily a list of literals. The additional field `nodupkey` is
+included to ensure that `not_tautology` is provable (which is needed to prove `insertRup_entails_hsat`
+and `insertRat_entails_hsat` in `LRAT.Formula.RupAddSound.lean` and `LRAT.Formula.RatAddSound.lean`).
+The additional field `nodup` is included to ensure that `delete` can be implemented by simply calling
+`erase` on the `clause` field. Without `nodup`, it would be necessary to iterate through the entire
+`clause` field and erase all instances of the literal to be deleted, since there would potentially
+be more than one.
+
+In principle, one could combine `nodupkey` and `nodup` to instead have one additional field that
+stipulates that `∀ l1 : PosFin numVarsSucc, ∀ l2 : PosFin numVarsSucc, l1.1 ≠ l2.1`. This would work
+just as well, and the only reason that `DefaultClause` is structured in this manner is that the
+`nodup` field was only included in a later stage of the verification process when it became clear
+that it was needed.
+-/
+@[ext] structure DefaultClause (numVarsSucc : Nat) where
+  clause : CNF.Clause (PosFin numVarsSucc)
+  nodupkey : ∀ l : PosFin numVarsSucc, (l, true) ∉ clause ∨ (l, false) ∉ clause
+  nodup : List.Nodup clause
+  deriving BEq
+
+instance : ToString (DefaultClause n) where
+  toString c := s!"{c.clause.reverse}"
+
+namespace DefaultClause
+
+def toList (c : DefaultClause n) : CNF.Clause (PosFin n) := c.clause
+
+theorem not_tautology (c : DefaultClause n) (l : Literal (PosFin n)) :
+    ¬ l ∈ toList c ∨ ¬Literal.negate l ∈ toList c := by
```

Comment:
Inconsistent spacing around the not here, I think the usual style is to omit the space.